<a href="https://colab.research.google.com/github/CoolerKula/Getting-Started-with-Text-Analysis-Project/blob/main/Getting_Started_with_Text_Analysis_Bernard_Kula.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<font color="#4b76b7">To start practicing, you will need to make a copy of it. Go to File > Save a Copy in Drive. You can then use the new copy that will appear in the new tab.</font>


# AfterWork Data Science: Getting Started with NLP Project

### Prerequisites

In [ ]:
# Importing the required libraries
# ---
# 
import pandas as pd # library for data manipulation
import numpy as np  # librariy for scientific computations
import re           # regex library to perform text preprocessing
import string       # library to work with strings
import nltk         # library for natural language processing
import scipy        # scientific conputing 

### 1. Importing our Data

In [ ]:
# Question: Given a new tweets, create a sentiment analysis model that will 
# predict whether a tweet will contain positive or negative sentiment.
# ---
# Dataset url = https://bit.ly/31kqByD 
# ---
#
df = pd.read_csv('https://bit.ly/31kqByD', encoding='latin-1')
df.head()

Unnamed: 0  ...  @switchfoot http://twitpic.com/2y1zl - Awww, that's a bummer.  You shoulda got David Carr of Third Day to do it. ;D
0      346508  ...  Obama forges his Muslim alliance against the c...                                                                  
1      883537  ...  Had the most spectacular prom ever  but now my...                                                                  
2      764173  ...  I am overwhelmed today  taking a moment to eat...                                                                  
3      638701  ...  @lindork Tres sad. I was totally a Max fan.  #...                                                                  
4      664821  ...  Crap, I was counting down the hours until my d...                                                                  

[5 rows x 7 columns]

### 2. Data Exploration

In [ ]:
# We can determine the size of our dataset
# ---
#
df.shape

(10000, 7)

Seems this dataset will need some data cleaning i.e. columns. We also don't need some columns to perform create our model. We will drop those columns.

### 3. Data Preparation

#### Basic Data Cleaning Techniques

In [ ]:
# We rename the columns for ease of referencing our columns later on
# ---
#
df.columns = ['id', 'target', 't_id', 'created_at', 'query', 'user', 'text']
df.head()

id  ...                                               text
0  346508  ...  Obama forges his Muslim alliance against the c...
1  883537  ...  Had the most spectacular prom ever  but now my...
2  764173  ...  I am overwhelmed today  taking a moment to eat...
3  638701  ...  @lindork Tres sad. I was totally a Max fan.  #...
4  664821  ...  Crap, I was counting down the hours until my d...

[5 rows x 7 columns]

In [ ]:
# We retain the relevant columns by dropping the columns we don't need 
# for creating a sentiment analysis model. 
# ---
#
df = df.drop(['id', 't_id', 'created_at', 'query', 'user'], axis = 1)
df.head()

target                                               text
0       0  Obama forges his Muslim alliance against the c...
1       4  Had the most spectacular prom ever  but now my...
2       0  I am overwhelmed today  taking a moment to eat...
3       0  @lindork Tres sad. I was totally a Max fan.  #...
4       0  Crap, I was counting down the hours until my d...

In [ ]:
# Understanding the distribution of target
# ---
#
df.target.value_counts() 

0    5067
4    4933
Name: target, dtype: int64

In [ ]:
# Let's determine whether our columns have the right data types
# ---
#
df.dtypes

target     int64
text      object
dtype: object

In [ ]:
# What values are in our target variable?
# ---
#
df.target.unique()

array([0, 4])

These are the two classes to which each document (text) belongs. The target value 0 means a text with a negative sentiment, while that of 4 means a text with a positive sentiment. 

In [ ]:
# Let's check for missing values 
# ---
# 
df.isnull().sum()

target    0
text      0
dtype: int64

We don't have any missing values, so we are good to go.

#### Text Processing

In [ ]:
# Text Cleaning: Removing all urls/links
# ---
# 
df['text'] =  df['text'].apply(lambda x: re.sub(r'http\S+|www\S+|https\S+','', str(x)))
df[['text']].head()

text
0  Obama forges his Muslim alliance against the c...
1  Had the most spectacular prom ever  but now my...
2  I am overwhelmed today  taking a moment to eat...
3  @lindork Tres sad. I was totally a Max fan.  #...
4  Crap, I was counting down the hours until my d...

In [ ]:
# Text Cleaning: Removing @ and # characters or replace them with space
# ---
# YOUR CODE GOES BELOW
#
df['text'] = df.text.str.replace('#','')
df['text'] = df.text.str.replace('@','')
df[['text']].sample(5)

text
8702  sleepysamco although I do miss English... grea...
8262  Holidays are nearly over and we all have 2go t...
9186  but it taught me a lesson, i learned something...
1466                                   I miss jn too   
4532  yelyahwilliams gotta love it when you leave cr...

In [ ]:
# Text Cleaning: Conversion to lowercase
# ---
# YOUR CODE GOES BELOW
#
df['text'] = df.text.apply(lambda x: " ".join(x.lower() for x in x.split()))
df['text'].sample(5)

4271    1400 dollars to fix the ac in my car... nooooo...
2786                             taking care of sick baby
9157    miabp hi5 for that! haha well don't talk about...
8232    just finished partying at 3 parties one night ...
6127    ok so there are pretty cool nuns one always ta...
Name: text, dtype: object

In [ ]:
# Text Cleaning: Splitting concatenated words
# ---
# Performing this step will take few minutes...
# ---
# YOUR CODE GOES BELOW
# 

# Installing wordnija and textblob
# ---
#
!pip3 install wordninja
!pip3 install textblob

# Importing those libraries
# ---
#
import wordninja 
from textblob import TextBlob

     |████████████████████████████████| 541 kB 14.4 MB/s 
  Created wheel for wordninja: filename=wordninja-2.0.0-py3-none-any.whl size=541551 sha256=7d79a9bf658df6b54039cabccea910167610adb9291b8ed78a7a061ff106e2b8
  Stored in directory: /root/.cache/pip/wheels/dd/3f/eb/a2692e3d2b9deb1487b09ba4967dd6920bd5032bfd9ff7acfc
Successfully built wordninja


In [ ]:
# Performing the split
# ---
#
df['text'] = df.text.apply(lambda x: wordninja.split(str(TextBlob(x))))  
df['text'] = df.text.str.join(' ')
df[['text']].sample(10) 

text
4058  kay le end uhh awesome im probably gonna stop ...
2672  crazy paws you probably think i need the extra...
2740                           shopping at the spectrum
8901            ok so it's definitely not so sunny lame
7184  one of those songs that make you realise how g...
3582  nick ol as writer you not seeing my picture or...
800   my s is just left for home wishing her a safe ...
754   tiffany eckhard t for women there is no such t...
8775       no frosty paws at the dog walk this year wtf
5178  was watching a movie and missed the opening of...

In [ ]:
# Text Cleaning: Removing punctuation characters
# ---
# YOUR CODE GOES BELOW
#
df['text'] = df.text.str.replace('[^\w\s]','') 

In [ ]:
# Text Cleaning: Removing stop words
# ---
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
stop = stopwords.words('english')
# YOUR CODE GOES BELOW
# 
df['text'] = df.text.apply(lambda x: " ".join(x for x in x.split() if x not in stop))
df[['text']].sample(10)

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


text
1877                               still lot work catch
7744     wen rich caps miss maybe see guys tonight late
9801                     partying getting drinks system
3233                       starring cca get monday fair
3278  going watch tv miki fuse finishing science wor...
8621                          good picture ronny winter
4276                ebony 1075 hah hah aha far away bet
6680              ry ssa choice icing theres neapolitan
7016  wavy stalker please tell dame isnt rapping lem...
5956  ao scott pans whatever works letdown larry fun...

In [ ]:
# Text Cleaning: Lemmatization
# ---
# YOUR CODE GOES BELOW
#

# For lemmatization, we will need to download wordnet
#
nltk.download('wordnet')
from textblob import Word
# Lemmatizing our text
# ---
#
df['lemmatization'] = df.text.apply(lambda x: " ".join([Word(word).lemmatize() for word in x.split()])) 
df[['text', 'lemmatization']].sample(10)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.


text                                      lemmatization
5033  tek chic laurie woohoo go laurie go tweet pair...  tek chic laurie woohoo go laurie go tweet pair...
8930  chris brogan thank hope great week ahead safe ...  chris brogan thank hope great week ahead safe ...
6052  amy isabella gonna happen send everyone love e...  amy isabella gonna happen send everyone love e...
1344                       britt deez lol youre excused                       britt deez lol youre excused
5629  jonas kevin awesome would love follow around w...  jonas kevin awesome would love follow around w...
3978                                    hate car drives                                     hate car drive
237                          g rum mi much fun one play                         g rum mi much fun one play
6309  one half hour phone con vos best friends good ...  one half hour phone con vos best friend good e...
6905  check mate zw ols show quo dave life quo sbs s...  check mate zw ols show quo dave life quo sb st...
9999       love 2 cook pie saw division 68 th didnt see       love 2 cook pie saw division 68 th didnt see

We won't remove numerics because we could loose meaning of our text if we lost the numerics. We could also further prepare our text by performing spelling correction but this is a resource intensive process that we will skip for now.

#### Feature Engineering Techniques 

In [ ]:
def get_polarity(text):
    try:
        textblob = TextBlob(unicode(text, 'utf-8'))
        pol = textblob.sentiment.polarity
    except:
        pol = 0.0
    return pol

In [ ]:
def get_subjectivity(text):
    try:
        textblob = TextBlob(unicode(text, 'utf-8'))
        subj = textblob.sentiment.subjectivity
    except:
        subj = 0.0
    return subj

In [ ]:
def avg_word(sentence):
  words = sentence.split()
  try:
    z = (sum(len(word) for word in words)/len(words))
  except ZeroDivisionError:
    z = 0 
  return z

In [ ]:
def pos_check(x, flag):
    cnt = 0
    try:
        wiki = TextBlob(x)
        for tup in wiki.tags:
            ppo = list(tup)[1]
            if ppo in pos_dic[flag]:
                cnt += 1
    except:
        pass
    return cnt

In [ ]:
# Feature Construction: Length of tweet
# ---

# YOUR CODE GOES BELOW
#
def feature_engineering(tweet):
  # Length of tweet
  df['length_of_tweet'] = df.text.str.len()

In [ ]:
# Feature Construction: Word count 
# ---
# YOUR CODE GOES BELOW
# 
df['word_count'] = df.text.apply(lambda x: len(str(x).split(" ")))

In [ ]:
# Feature Construction: Word density (Average no. of words / tweet)
# ---
# YOUR CODE GOES BELOW
#
df['avg_word_length'] = df.text.apply(lambda x: avg_word(x))

In [ ]:
# Feature Construction: Noun count
# ---
# YOUR CODE GOES BELOW
#
# First, we will download the punkt and the averaged_perceptron_tagger into our notebook environment. 
# which will allow us to find the part of speech tags.
# ---
#
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')

# We create the function to check and get the part of speech tag count of a words in a given sentence
pos_dic = {
    'noun' : ['NN','NNS','NNP','NNPS'],
    'pron' : ['PRP','PRP$','WP','WP$'],
    'verb' : ['VB','VBD','VBG','VBN','VBP','VBZ'],
    'adj' :  ['JJ','JJR','JJS'],
    'adv' : ['RB','RBR','RBS','WRB']
}

def pos_check(x, flag):
    cnt = 0
    try:
        wiki = TextBlob(x)
        for tup in wiki.tags:
            ppo = list(tup)[1]
            if ppo in pos_dic[flag]:
                cnt += 1
    except:
        pass
    return cnt

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


In [ ]:
# Noun Count
# ---
# YOUR CODE GOES BELOW
#
df['noun_count'] = df.text.apply(lambda x: pos_check(x, 'noun'))

In [ ]:
# Feature Construction: Verb count
# ---
# YOUR CODE GOES BELOW
#
df['verb_count'] = df.text.apply(lambda x: pos_check(x, 'verb'))

In [ ]:
# Feature Construction: Adjective count / Tweet
# ---
# YOUR CODE GOES BELOW
#
df['adj_count'] = df.text.apply(lambda x: pos_check(x, 'adj'))

In [ ]:
# Feature Construction: Adverb count / Tweet
# ---
# YOUR CODE GOES BELOW
#
df['adv_count'] = df.text.apply(lambda x: pos_check(x, 'adv'))

In [ ]:
# Feature Construction: Pronoun 
# ---
# YOUR CODE GOES BELOW
#
df['pron_count'] = df.text.apply(lambda x: pos_check(x, 'pron'))

In [ ]:
# Feature Construction: Subjectivity
# ---
# YOUR CODE GOES BELOW
# 
df['subjectivity'] = df.text.apply(get_subjectivity)

In [ ]:
# Feature Construction: Polarity
# ---
# YOUR CODE GOES BELOW
# 
df['polarity'] = df.text.apply(get_polarity)

In [ ]:
# Feature Construction: Word Level N-Gram TF-IDF Feature 
# ---
# YOUR CODE GOES BELOW
#
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf = TfidfVectorizer(max_features=1000, lowercase=True, analyzer='word', ngram_range=(1,3),  stop_words= 'english')
df_word_vect = tfidf.fit_transform(df.text)

In [ ]:
# Feature Construction: Character Level N-Gram TF-IDF Feature
# ---
# YOUR CODE GOES BELOW
# 
tfidf = TfidfVectorizer(max_features=1000, lowercase=True, analyzer='char', ngram_range=(1,3),  stop_words= 'english')
df_char_vect = tfidf.fit_transform(df.text)

In [ ]:
# Let's prepare the constructed features for modeling
# ---
#
X_metadata = np.array(df.iloc[:, 3:12])
X_metadata

array([[11.        ,  5.27272727,  6.        , ...,  0.        ,
         0.        ,  0.        ],
       [12.        ,  6.        ,  5.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 6.        ,  5.83333333,  4.        , ...,  0.        ,
         0.        ,  0.        ],
       ...,
       [ 8.        ,  4.75      ,  3.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 6.        ,  4.83333333,  2.        , ...,  0.        ,
         0.        ,  0.        ],
       [10.        ,  3.5       ,  5.        , ...,  0.        ,
         0.        ,  0.        ]])

In [ ]:
# We combine our two tfidf (sparse) matrices and X_metadata
# ---
#
X = scipy.sparse.hstack([df_word_vect, df_char_vect, X_metadata])
X

<10000x2009 sparse matrix of type '<class 'numpy.float64'>'
	with 933618 stored elements in COOrdinate format>

In [ ]:
# Getting our response variable
# ---
#
y = np.array(df.iloc[:, 0])
y

array([0, 4, 0, ..., 0, 4, 0])

### 4. Data Modelling

During this step, we will use machine learning algorithms to train and test our sentiment analysis models.

In [ ]:
# Splitting our data
# ---
#
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
# Fitting our model
# ---
#

# Importing the algorithms
from sklearn.naive_bayes import MultinomialNB 
from sklearn.linear_model import LogisticRegression

nb_classifier = MultinomialNB() 
lr_classifier = LogisticRegression(max_iter=1000) 

# Training our model
nb_classifier.fit(X_train, y_train) 
lr_classifier.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [ ]:
# Making predictions
# ---
#
y_predict_nb = nb_classifier.predict(X_test) 
y_predict_lr = lr_classifier.predict(X_test)

In [ ]:
# Evaluating the Models
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Accuracy scores
# ---
#
print("Naive Bayes Classifier:\n", accuracy_score(y_test, y_predict_nb)) 
print("Logistic Regression Classifier: \n", accuracy_score(y_test, y_predict_lr))

Naive Bayes Classifier:
 0.726
Logistic Regression Classifier: 
 0.728


In [ ]:
# Confusion matrices
# ---
# 
print("Naive Bayes Classifier: \n", confusion_matrix(y_test, y_predict_nb)) 
print("Logistic Regression Classifier: \n", confusion_matrix(y_test, y_predict_lr))

Naive Bayes Classifier: 
 [[756 294]
 [254 696]]
Logistic Regression Classifier: 
 [[759 291]
 [253 697]]


In [ ]:
# Classification Reports
# ---
#
print("Naive Bayes Classifier: \n", classification_report(y_test, y_predict_nb)) 
print("Logistic Regression Classifier: \n", classification_report(y_test, y_predict_lr))

Naive Bayes Classifier: 
               precision    recall  f1-score   support

           0       0.75      0.72      0.73      1050
           4       0.70      0.73      0.72       950

    accuracy                           0.73      2000
   macro avg       0.73      0.73      0.73      2000
weighted avg       0.73      0.73      0.73      2000

Logistic Regression Classifier: 
               precision    recall  f1-score   support

           0       0.75      0.72      0.74      1050
           4       0.71      0.73      0.72       950

    accuracy                           0.73      2000
   macro avg       0.73      0.73      0.73      2000
weighted avg       0.73      0.73      0.73      2000



**Evaluation our Models**

* **Accuracy:** the percentage of texts that were assigned the correct topic.
* **Precision:** the percentage of texts the classifier classified correctly out of the total number of texts it predicted for each topic
* **Recall:** the percentage of texts the model predicted for each topic out of the total number of texts it should have predicted for that topic.
* **F1 Score:** the average of both precision and recall.

To improve our model, we can try perfoming other text processing techniques that would better prepare our data for fitting our model. We can also use different vectorizing techniques, implement other machine learning models and perform hyperparameter tuning.

### 5. Recommendations


Our best model had an accuracy of 72.8% and thus we can use it for classifying newer tweets. We can improve this performance by performing hyperparameter tuning and feature engineering methods. 